### ISSUE TITLE
Atomic signal masks for reliably interruptible device I/O

### ISSUE THREAD

**Issue by riastradh** [Yubico/libfido2#251]
**Describe the feature you would like.**
I would like to transact asynchronously with the first of any (reasonable) number of authenticator devices in the system, like https://github.com/mozilla/authenticator-rs does.  One way to do this would be to completely redesign the API to be asynchronous and event-driven (and it would also be nice to wait for the user to insert the device, by waiting for udev/drvctl/&c. events).  But that's obviously a lot of work.

Short of that, querying the devices in separate pthreads will serve as an acceptable alternative: according to https://github.com/Yubico/libfido2/issues/43, libfido2 generally allows operations on different objects in different threads as long as the threads themselves serialize the access, and there should usually be few enough devices that a profusion of threads is not an issue.  Rough sketch (omitting details of locking, error checking, and so on):

```
/* Main thread */
pthread_create(worker, devinfo1);
pthread_create(worker, devinfo2);
...
while (!done) {
	if (pthread_cond_timedwait(...) == ETIMEDOUT)
		break;
}

/* Worker N */
devN = fido_dev_new();
fido_dev_open(devN, devinfoN);
fido_dev_make_cred(devN, credN);
done = 1;
pthread_cond_signal(...);
```

But there's a snag.  Once the _first_ device has answered a transaction such as `fido_dev_make_cred`, the other transactions and threads need to be interrupted so they don't continue to hang waiting for input that's not coming and would be discarded anyway.  Two obvious approaches are:

1. Set up a null signal handler, and have the main thread deliver SIGUSR1 to all the worker threads that haven't completed yet, so that (e.g.) `poll` in `fido_hid_unix_wait` will return early with EINTR.
2. When a worker thread completes, have the main thread issue `fido_dev_cancel` for all devices that haven't returned yet, so that (e.g.) the CBOR authenticatorMakeCredential transaction will abort with error.

However, both of these approaches necessarily have race conditions, because there's no way for the main thread to know exactly when the CTAP transaction has actually begun: deliver a signal too early, and it won't interrupt the `poll` inside libfido2 because `poll` hasn't started yet; cancel a transaction too early, and it won't interrupt the pending transaction because the transaction isn't pending yet.  Either way, the worker thread will hang when it should respond promptly to interruption.

This could be fixed by passing a signal mask for `poll` (or rather, `ppoll`) to set atomically when it is actually ready to be interrupted.  One way would be to pass it in to all the routines that may block: `fido_dev_open`, `fido_dev_make_cred`, `fido_dev_get_assert`, &c., either by adding arguments or adding new symbols, but that's not great for compatibility or simplicity.  Another way would be to add `fido_dev_sigmask(dev, sigmask)` -- this would have the effect that any such blocking routines (`fido_dev_make_cred`, &c.) would atomically set the signal mask to `sigmask` when they're about to block.

```
/* Main thread, continued */
pthread_kill(thread1, SIGUSR1);
pthread_kill(thread2, SIGUSR1);
...

/* Worker N */
sigset_t blockusr1, unblockall;
sigemptyset(&blockusr1);
sigaddset(&blockusr1, SIGUSR1);
sigemptyset(&unblockall);
pthread_sigmask(SIG_SETMASK, &blockusr1);
if (done)
	/* back out */
devN = fido_dev_new();
fido_dev_sigmask(devN, &unblockall);
if (fido_dev_open(devN, pathN) != FIDO_OK ||
    fido_dev_make_cred(devN, credN) != FIDO_OK) {
	fido_dev_cancel(devN);
	/* back out */
}
done = 1;
pthread_cond_signal(...);
```

Internally, this would require minor changes to all the hid_* drivers to pass the signal mask through into the blocking routine, most of which can just defer to `fido_hid_unix_wait` to use `ppoll`.  It may also be necessary to wait for writes to be ready with `ppoll` too, not just reads.

**Describe the impact to existing functionality.**
Existing functionality should be unaffected -- the semantics only changes if you use the proposed new interface.

**Describe how you are willing to help.**
I can draft a patch to do this on NetBSD; the semantics should be straightforward to carry over to other operating systems.

**Comment by martelletto** [comment]
Thank you for the elaborate feature request. The suggested approach makes sense to me.

**Comment by riastradh** [comment]
I drafted an implementation of this proposal for consideration: https://github.com/Yubico/libfido2/pull/252, tested with the attached [test.c](https://github.com/Yubico/libfido2/files/5727112/test.c.txt) on NetBSD.

I don't know what libfido2's compatibility rules are -- if you're willing to require clients to recompile, the change as is might be OK (once all the other hid_* back ends are adapted); otherwise it will be necessary to find a way to make `fido_dev_set_io_functions` and `fido_dev_set_transport_functions` work, perhaps by introducing new symbols like `fido_dev_set_io_functions_v2` and `fido_dev_set_transport_functions_v2` and leaving in logic to handle non-sigmasked I/O and transport functions for programs that don't need reliable interruption.

What do you think?

**Comment by martelletto** [comment]
Thank you for the PR. I would suggest moving the signal mask inside `struct hid_netbsd`, and renaming `fido_dev_sigmask()` to `fido_dev_set_sigmask()` for consistency with the rest of the API. `fido_dev_set_sigmask()` could then delegate the task to a corresponding `fido_hid_set_sigmask()` in the HID backend.

This would allow us to keep the I/O and transport functions as they are, at the expense of losing access to the signal mask in `fido_hid_open()`. Since `fido_hid_open()` only uses the signal mask to run `terrible_ping_kludge()` -- which shouldn't exist in the first place -- that seems acceptable to me. Does that make sense?

**Comment by martelletto** [comment]
On a side note regarding test.c, do you plan to support the case where a PIN is needed for `fido_dev_make_cred()` to succeed (e.g. a FIDO2 token with a PIN set)?

**Comment by riastradh** [comment]
> Thank you for the PR. I would suggest moving the signal mask inside `struct hid_netbsd`, and renaming `fido_dev_sigmask()` to `fido_dev_set_sigmask()` for consistency with the rest of the API. `fido_dev_set_sigmask()` could then delegate the task to a corresponding `fido_hid_set_sigmask()` in the HID backend.
> 
> This would allow us to keep the I/O and transport functions as they are, at the expense of losing access to the signal mask in `fido_hid_open()`. Since `fido_hid_open()` only uses the signal mask to run `terrible_ping_kludge()` -- which shouldn't exist in the first place -- that seems acceptable to me. Does that make sense?

Yes, that makes sense.  Done in a new PR https://github.com/Yubico/libfido2/pull/254 for comparison.  How does that look?

Updated [test_sigmask.c.txt](https://github.com/Yubico/libfido2/files/5730977/test_sigmask.c.txt) program, since fido_dev_set_sigmask must happen after fido_dev_open now; I also simplified some of the logic.

> `terrible_ping_kludge()` -- which shouldn't exist in the first place

I agree.  I have a draft fix in the NetBSD kernel USB stack.  Unfortunately I'm not sure I see any better way to work around it from userland until NetBSD 9.1 is EOL -- from what I can tell by a cursory reading of the USB spec, resynchronizing the data toggle requires the host to issue ClearFeature(ENDPOINT_HALT), and I don't see an existing path to do that.

> On a side note regarding test.c, do you plan to support the case where a PIN is needed for `fido_dev_make_cred()` to succeed (e.g. a FIDO2 token with a PIN set)?

I hadn't thought about that, no -- PINs aren't really on my radar; my interest is mainly in simple devices that just work out of the box with no configuration or management tools whatsoever.  That said, I assume the same mechanism could be made to work with a blocking variant of fido_dev_get_touch_begin/status, which would also take advantage of the signal mask, rather than busy-waiting on the host.

**Comment by martelletto** [comment]
Thanks. I took the liberty of adding a commit to the PR; hope you won't mind. It is unfortunate that MSVC does not define `sigset_t`, so we can't safely define `fido_dev_set_sigmask()` on Windows (even as a no-op) unless we have it take a `fido_sigset_t`.

**Comment by riastradh** [comment]
> Thanks. I took the liberty of adding a commit to the PR; hope you won't mind.

Of course!  (I should rebase it to remove `WIP:` from the commit message before it gets merged, I suppose.)

I note that if libfido2 calls the OS-dependent fido_hid_set_sigmask directly, then it won't work with applications that use fido_dev_set_io_functions.  But maybe that's not a big deal -- such applications can, of course, use whatever signal mask they see fit inside the io functions; it just strikes me as a little confusing for composition.

> It is unfortunate that MSVC does not define `sigset_t`, so we can't safely define `fido_dev_set_sigmask()` on Windows (even as a no-op) unless we have it take a `fido_sigset_t`.

Well, I guess Windows applications wouldn't have any occasion to use it anyway.

**Comment by riastradh** [comment]
Re -std=c99: perhaps it would be better to define _POSIX_C_SOURCE to be 200809L than to remove -std=c99 altogether?  I would expect that should expose the relevant types and declarations (but I haven't tested).

**Comment by riastradh** [comment]
I pushed [another commit](https://github.com/Yubico/libfido2/pull/254/commits/40cb69a17d8e23bdd92ba5c568a986e10e5c767a) that defines _POSIX_C_SOURCE to 200809L globally, as well as defining _XOPEN_SOURCE, _XOPEN_SOURCE_EXTENDED, and/or _NETBSD_SOURCE in a handful of files that use non-POSIX extensions.  This more clearly identifies the standard libfido2 generally relies on, and marks the use of nonstandard/nonportable extensions, which enables more immediate feedback from the compiler about where such extensions are inadvertently being used.

Obviously it's up to you to decide whether you want to go this way or just enable all the nonstandard extensions all the time, which is a little easier in the short term but more likely to cause headaches downstream -- feel free to discard that commit or tweak it as you see fit.

**Comment by riastradh** [comment]
There's a problem with conditionalizing the declaration of fido_dev_set_sigmask on HAVE_SIGSET_T in the public fido.h header file: HAVE_SIGSET_T is defined internally to the libfido2 build, but it is not necessarily defined by clients that simply use libfido2.  Perhaps it should be conditional on _POSIX_C_SOURCE or !_MSVC or something instead?

**Comment by martelletto** [comment]
> I pushed [another commit](https://github.com/Yubico/libfido2/pull/254/commits/40cb69a17d8e23bdd92ba5c568a986e10e5c767a) that defines _POSIX_C_SOURCE to 200809L globally, as well as defining _XOPEN_SOURCE, _XOPEN_SOURCE_EXTENDED, and/or _NETBSD_SOURCE in a handful of files that use non-POSIX extensions. This more clearly identifies the standard libfido2 generally relies on, and marks the use of nonstandard/nonportable extensions, which enables more immediate feedback from the compiler about where such extensions are inadvertently being used.

Thank you. I agree that's a better approach. I had a brief look on macOS; combining `_POSIX_C_SOURCE=200809L` with `-std=c99` has the opposite effect it does on Linux, and we end up having problems with newish additions like `timingsafe_bcmp()`. Getting the nuances right for the various platforms will take more time than I have at the moment, so I would prefer to implement this approach on a separate PR if possible.

**Comment by martelletto** [comment]
> There's a problem with conditionalizing the declaration of fido_dev_set_sigmask on HAVE_SIGSET_T in the public fido.h header file: HAVE_SIGSET_T is defined internally to the libfido2 build, but it is not necessarily defined by clients that simply use libfido2. Perhaps it should be conditional on _POSIX_C_SOURCE or !_MSVC or something instead?

Indeed; thank you so much. I have pushed a branch with a few additions and minor modifications: https://github.com/Yubico/libfido2/tree/sigmask. I have also rebased the branch on top of HEAD. Please feel free to cherry-pick, squash, etc. as you see fit, and season's greetings!

**Comment by riastradh** [comment]
Thanks, I rebased to fix the commit message, verified that it still works, and created a replacement PR: https://github.com/Yubico/libfido2/pull/255

Feel free to merge as is, or squash and merge, or whatever.
